### PR TITLE
Localization of msre_edit.php

### PR DESCRIPTION
**Pull request by stefaweb**
Localization of msre_edit.php, msre_index.php, misrule.php and msre_table_functions.php.
 
Functions to move in functions.php
See issue #624 for bug.

I don't like the system to "reload" MailScanner. Do we really needed that as MS reload itself regularly? Do we have another way to do this?

### CODE REVIEW

File: mailscanner/msre_edit.php [hunk 1]
```diff
@@ -317,14 +317,14 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
         $desc_value = htmlentities($desc, ENT_QUOTES);
         if (preg_match('/#DISABLED#/', $rule_part['0direction'])) {
             $rule_disabled = 1;
-            $rule_action_select_options = "<option value=\"Enable\">Enable</option>\n";
+            $rule_action_select_options = '<option value=\"Enable\">' . __('enable55') . '</option>' . "\n";
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: when using single quote, there is no more need to escape double quote:
`$rule_action_select_options = '<option value="Enable">' . __('enable55') . '</option>' . "\n";`

Comment (stefaweb) [inline mailscanner/msre_edit.php]: Ok will do.
I've to stop for today.
Look some parts have path hard coded.

**Comment by stefaweb** [comment]
Localisation done.
Need de.php

**Comment by Skywalker-11** [comment]
de.php

```
'msreedit55' => 'Regelsatz für MailScanner bearbeiten',
    'submit55' => 'Submit',
    'enable55' => 'Aktivieren',
    'disable55' => 'Deaktivieren',
    'description55' => 'Beschreibung:',
    'action55' => 'Aktion:',
    'savevalue55' => 'Änderungen speichern',
    'backmsre55' => 'Zurück zum Verzeichnis der MSRE Regelsätze',
    'backmw55' => 'Zurück zu MailWatch:',
    'schedureloadmw55' => 'Neustart von MailScanner wird geplant ...',
    'error0155' => 'FEHLER: Neustart von MailScanner konnte nicht geplant werden! (Sie müssen MailScanner manuell über die Kommandozeile neu starten)',
    'ok55' => 'Ok',
    'message55' => 'Ihre Änderungen werden in den nächsten %s Minuten angewendet, sobald MailScanner neu gestartet wird.',
    'backupfile55' => 'Sichern der aktuellen Datei...',
    'error0255' => 'FEHLER: Es konnte keine Sicherung angelegt werden!',
    'error0355' => 'FEHLER: Konnte %s nicht zum Schreiben öffnen !',
    'contentsof55' => 'Momentaner Inhalt von %s:',
    'editrules55' => 'Regelsatz bearbeiten:',
    'openwriting55' => 'Öffne %s zum Schreiben...',
    'writefile55' => 'Schreibe neue Datei...',
    'writebytes55' => '%s Bytes geschrieben.',
    'fileclosed55' => 'Datei geschlossen.',
    'donewrite55' => 'Schreiben der Datei abgeschlossen.',
```

**Comment by stefaweb** [comment]
More de.php to do in 29-msre_index.php and 30-msrule.php

File: mailscanner/msre_edit.php [hunk 2]
```diff
@@ -291,79 +257,83 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
-            $rule_action_select_options = "<option value=\"Enable\">Enable</option>\n";
+            $rule_action_select_options = '<option value="Enable">' . __('enable55') . '</option>' . "\n";
             $disable_desc_text = ' disabled ';
             $desc_field = 'rule' . $rule_count . '_description_disabled';
             $hidden_field_code = '<input type="hidden" name="' .
                 preg_replace('/_disabled$/', '', $desc_field) . "\" value=\"$desc_value\">";
         } else {
             $rule_disabled = 0;
-            $rule_action_select_options = "<option value=\"Disable\">Disable</option>\n";
+            $rule_action_select_options = '<option value=\"Disable\">' . __('disable55') .'</option>' . "\n";
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: remove escape on double quote

File: mailscanner/msre_edit.php [hunk 3]
```diff
@@ -291,79 +257,83 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
-        // now continue on..
-        $rule_action_select_html .= ">\n" .
-            "  <option value=\"\" selected>----</option>\n" .
-            "  <option value=\"Delete\">Delete</option>\n" .
-            $rule_action_select_options . "</select>\n";
+        
+        // Now continue on.
+        $rule_action_select_html .= '>\n' .
+            '  <option value="" selected>----</option>' . "\n" .
+            '  <option value=\Delete">Delete</option>' . "\n" .
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: missing double quote (and extra \)

File: mailscanner/msre_edit.php [hunk 4]
```diff
@@ -291,79 +257,83 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
+            $rule_action_select_options . '</select>' . "\n";
         $desc_text = array(
             $rule_action_select_html => 'rowspan="3"',
-            '<b>Description:</b>&nbsp;&nbsp;<input type="text" ' .
-            "name=\"$desc_field\" size=\"95\" value=\"" . $desc_value . '"' .
+            '<b>' . __('description55') . '</b>&nbsp;&nbsp;<input type="text" ' .
+            'name="$desc_field" size="95" value="' . $desc_value . '"' .
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: $desc_field values is not printed

File: mailscanner/msre_edit.php [hunk 5]
```diff
@@ -291,79 +257,83 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
             switch (strtolower($key)) {
                 case '2and':
-                    // and gets a checkbox
+                    // And gets a checkbox
                     $field_name = $part_name;
                     if ($rule_disabled) {
                         $field_name .= '_disabled';
                     }
-                    $checkbox_html = "<input type=\"checkbox\" name=\"$field_name\" value=\"";
+                    $checkbox_html = '<input type="checkbox" name="$field_name" value="';
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: $field_name value is not printed

File: mailscanner/msre_edit.php [hunk 6]
```diff
@@ -374,29 +344,29 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
                     }
                     $checkbox_html .= '>and';
                     if ($rule_disabled) {
-                        $checkbox_html .= "\n<input type=\"hidden\" name=\"$part_name\" value=\"";
+                        $checkbox_html .= '\n<input type="hidden" name="$part_name" value="';
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: `\n` must be in double quote, $part_name value is not printed

File: mailscanner/msre_edit.php [hunk 7]
```diff
@@ -374,29 +344,29 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
                         if ($value) {
                             $checkbox_html .= 'and';
                         }
-                        $checkbox_html .= "\">\n";
+                        $checkbox_html .= '">\n';
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: `\n` must be in double quote

File: mailscanner/msre_edit.php [hunk 8]
```diff
@@ -374,29 +344,29 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
                 case '0direction':
                 case '3and_direction':
-                    // these get select boxen
+                    // These get select boxen
                     $field_name = $part_name;
                     if ($rule_disabled) {
                         $field_name .= '_disabled';
                     }
-                    $select_html = "<select name=\"$field_name\"";
+                    $select_html = '<select name="$field_name"';
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: $field_name value is not printed

File: mailscanner/msre_edit.php [hunk 9]
```diff
@@ -374,29 +344,29 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
                     if ($rule_disabled) {
                         $select_html .= ' disabled ';
                     }
-                    $select_html .= ">\n" .
+                    $select_html .= '>\n' .
                         '<option value=""></option>';
                     foreach ($CONF_ruleset_keyword as $current_kw) {
-                        $select_html .= "<option value=\"$current_kw\"";
+                        $select_html .= '<option value="$current_kw"';
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: $current_kw  value is not printed

File: mailscanner/msre_edit.php [hunk 10]
```diff
@@ -374,29 +344,29 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
                     if ($rule_disabled) {
                         $field_name .= '_disabled';
                     }
-                    $select_html = "<select name=\"$field_name\"";
+                    $select_html = '<select name="$field_name"';
                     if ($rule_disabled) {
                         $select_html .= ' disabled ';
                     }
-                    $select_html .= ">\n" .
+                    $select_html .= '>\n' .
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: `\n` must be in double quote

File: mailscanner/msre_edit.php [hunk 11]
```diff
@@ -414,33 +384,33 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
                         if (strtolower($current_kw) === $kw) {
                             $select_html .= ' selected';
                         }
-                        $select_html .= ">$current_kw</option>";
+                        $select_html .= '>$current_kw</option>';
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: $current_kw value is not printed

File: mailscanner/msre_edit.php [hunk 12]
```diff
@@ -414,33 +384,33 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
                     }
-                    // need to close my select tag..
+                    // Need to close my select tag..
                     $select_html .= '</select>';
 
                     if ($rule_disabled) {
-                        $select_html .= "\n<input type=\"hidden\" name=\"$part_name\" value=\"$value\">\n";
+                        $select_html .= '\n<input type="hidden" name="$part_name" value="$value">\n';
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: $part_name and $value  values are not printed
`\n` must be in double quote

File: mailscanner/msre_edit.php [hunk 13]
```diff
@@ -414,33 +384,33 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
                 default:
-                    // others get regular text boxen
+                    // Others get regular text boxen
                     $field_name = $part_name;
                     if ($rule_disabled) {
                         $field_name .= '_disabled';
                     }
                     if (strtolower($key) === '99action') {
-                        $temp_text = '</td></tr><tr><td colspan="' . (MSRE_COLUMNS - 1) . "\"><b>Action:</b>&nbsp;&nbsp;<input type=\"text\" name=\"$field_name\" value=\"$value\" size=\"100\"";
+                        $temp_text = '</td></tr><tr><td colspan="' . (MSRE_COLUMNS - 1) . '"><b>' . __('action55') . '</b>&nbsp;&nbsp;<input type="text" name="$field_name" value="$value" size="100"';
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: $field_name and $value values are not printed

File: mailscanner/msre_edit.php [hunk 14]
```diff
@@ -414,33 +384,33 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
                     } else {
-                        $temp_text = "<input type=\"text\" name=\"$field_name\" value=\"$value\"";
+                        $temp_text = '<input type="text" name="$field_name" value="$value"';
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: $field_name and $value values are not printed

File: mailscanner/msre_edit.php [hunk 15]
```diff
@@ -414,33 +384,33 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
                     }
                     if ($rule_disabled || (strtolower($key) === '1target' && strtolower($value) === 'default')) {
                         $temp_text .= ' disabled ';
                     }
                     $temp_text .= '>';
                     if ($rule_disabled) {
-                        $temp_text .= "\n<input type=\"hidden\" name=\"$part_name\" value=\"$value\">\n";
+                        $temp_text .= '\n<input type="hidden" name="$part_name" value="$value">' . "\n";
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: $part_name and $value values are not printed
`\n` in double quote

File: mailscanner/msre_edit.php [hunk 16]
```diff
@@ -457,90 +427,87 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
-        // something new for v0.2.1 ... I'm going to try to put a box
+        // Something new for v0.2.1. I'm going to try to put a box
         // around each ruleset, so it's easier to pick them out from
-        // each other in the list...
-        echo "<tr>\n" .
-            "<td class=\"$boxclass\" colspan=\"" . MSRE_COLUMNS . "\">\n" .
-            "<table border=\"0\">\n";
+        // each other in the list.
+        echo '<tr>' . "\n" .
+            '<td class="$boxclass" colspan="' . MSRE_COLUMNS . '>' . "\n" . '<table border="0">' . "\n";
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: $boxclass value is not printed

File: mailscanner/msre_edit.php [hunk 17]
```diff
@@ -457,90 +427,87 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
         'target" value="">';
-    // and
     $add_rule_text[] = '<input type="checkbox" name="' . $add_prefix .
         'and" value="and">and';
-    // and direction
     $temp_html = '<select name="' . $add_prefix .
         'and_direction"><option value=""></option>';
     foreach ($CONF_ruleset_keyword as $kw) {
-        $temp_html .= "<option value=\"$kw\">$kw</option>";
+        $temp_html .= '<option value="$kw">$kw</option>';
```
Comment (endelwar) [inline mailscanner/msre_edit.php]: $kw values are not printed

**Comment by stefaweb** [comment]
Some words again to localize in DE in:
29-msre_index.php
30-msrule.php
55-msre_edit.php

File: mailscanner/languages/de.php [hunk 18]
```diff
@@ -614,9 +614,14 @@
...
     'rulesetedit29' => 'Regelsatz Editor',
     'auditlog29' => 'Anzeigeversuch eines Nicht-Admins der MailScanner-Seite zur Regelbearbeitung',
+    'editrule29' => 'Choose a ruleset to edit:',
+    'norulefound29' => 'No rules found',
 
     // 30-msrule.php
     'rules30' => 'Regeln',
+    'dirblocked30' => 'Directory traversal attempt blocked.',
+    'unableopenfile30' => 'Unable to open file.',
+    'file30' => 'File:',
```
Comment (Skywalker-11) [inline mailscanner/languages/de.php]:
```
'dirblocked30' => 'Durchsuchen der Verzeichnisse wurde blockiert.',
'unableopenfile30' => 'Datei konnte nicht geöffnet werden.',
'file30' => 'Datei:',
```

File: mailscanner/languages/de.php [hunk 19]
```diff
@@ -614,9 +614,14 @@
     // 29-msre_index.php
     'rulesetedit29' => 'Regelsatz Editor',
     'auditlog29' => 'Anzeigeversuch eines Nicht-Admins der MailScanner-Seite zur Regelbearbeitung',
+    'editrule29' => 'Choose a ruleset to edit:',
+    'norulefound29' => 'No rules found',
```
Comment (Skywalker-11) [inline mailscanner/languages/de.php]:
```
  'editrule29' => 'Wähle einen Regelsatz zum Bearbeiten:',
  'norulefound29' => 'Keine Regeln gefunden',
```

File: mailscanner/languages/de.php [hunk 20]
```diff
@@ -787,6 +792,30 @@
     // 55-msre_edit.php
     'diefnf55' => 'Datei nicht gefunden:',
     'auditlog55' => 'Nicht-Admin versuchte die MailScanner-Seite zur Regelbearbeitung anzuzeigen',
+    'msreedit55' => 'Regelsatz für MailScanner bearbeiten',
+    'submit55' => 'Submit',
```
Comment (Skywalker-11) [inline mailscanner/languages/de.php]:
```
   'submit55' => 'Senden',
```

File: mailscanner/languages/de.php [hunk 21]
```diff
@@ -787,6 +792,30 @@
...
+    'error0255' => 'FEHLER: Es konnte keine Sicherung angelegt werden!',
+    'error0355' => 'FEHLER: Konnte %s nicht zum Schreiben öffnen !',
+    'contentsof55' => 'Momentaner Inhalt von %s:',
+    'editrules55' => 'Regelsatz bearbeiten:',
+    'openwriting55' => 'Öffne %s zum Schreiben...',
+    'writefile55' => 'Schreibe neue Datei...',
+    'writebytes55' => '%s Bytes geschrieben.',
+    'fileclosed55' => 'Datei geschlossen.',
+    'donewrite55' => 'Schreiben der Datei abgeschlossen.',
+    'conditions55' => 'Condition:',
```
Comment (Skywalker-11) [inline mailscanner/languages/de.php]:
```
   'conditions55' => 'Bedingungen:',
```

File: mailscanner/msre_edit.php [hunk 22]
```diff
@@ -103,89 +81,77 @@
...
+    echo '<input type="hidden" name="token" value="' . $_session['token'] . '">' . "\n";
+    echo '<input type="hidden" name="formtoken" value="' . generateformtoken('/msre_edit.php form token') . '">' . "\n";
+    echo '<input type="submit" name="submit" value="' . __('submit55') . '">' . "\n";
+    echo '<input type="hidden" name="submitted" value="1">' . "\n";
+    
+    // Check for status message, and append it to the end of the header
     $my_header = '';
     if ($status_msg) {
-        $my_header .= "<br>\n" . $status_msg;
+        $my_header .= '<br>\n' . $status_msg;
```
Comment (Skywalker-11) [inline mailscanner/msre_edit.php]: the \n will be printed as line break but as the plain chars `\` and `n`. Has to be in double quotes or use PHP_EOL

File: mailscanner/msre_edit.php [hunk 23]
```diff
@@ -566,7 +533,7 @@ function Process_Form($file_contents, $short_filename)
     // stopping when we have reached a non-comment line
     $previous_line = '';
     $first_line = true;
-    foreach (preg_split("/\n/", $file_contents) as $line) {
+    foreach (preg_split('/\n/', $file_contents) as $line) {
```
Comment (Skywalker-11) [inline mailscanner/msre_edit.php]: no sure if this will work as it's not a line break anymore

Comment (stefaweb) [inline mailscanner/msre_edit.php]: J'ai fait toute la partie traduction et mise à jour de syntax. Ce serait mieux pour la partie code que quelqu'un d'autre reprenne le flambeau.

**Comment by stefaweb** [comment]
I tried the code. 

```
Notice: Undefined variable: _session in /var/www/html/mailscanner/msre_edit.php on line 132
```

**Comment by Skywalker-11** [comment]
should be `$_SESSION['token'] `

**Comment by stefaweb** [comment]
This doesn't work.
Current contents of %s: (contentsof55)

`TRH_Single(sprintf(__('contentsof55'), $short_filename), 'colspan="' . MSRE_COLUMNS . '"');`

All From, To, and others "and" need to be translated. Not easy. It's mixed in the code.

capture d ecran 2017-04-07 a 16 41 00

capture d ecran 2017-04-07 a 16 41 07

**Comment by stefaweb** [comment]
New to translate:

'and55' => 'and',

**Comment by stefaweb** [comment]
This one is not selectable?

capture d ecran 2017-04-07 a 16 57 30

**Comment by Skywalker-11** [comment]
de: `'and55' => 'und',`

**Comment by Skywalker-11** [comment]
> This one is not selectable?

Yes its a default that has to be there and should not be edited.

**Comment by stefaweb** [comment]
Page layout!!!!

capture d ecran 2017-04-07 a 17 08 24

**Comment by Skywalker-11** [comment]
There are two submit buttons 😄 

image

image

And you are right: you destroyed the bit of layout that was there 😁 
image

The cause is that all the input and selects are outside the table

File: mailscanner/msre_edit.php [hunk 24]
```diff
@@ -103,89 +81,77 @@
...
+    // Show page header
     if ($my_header) {
         TR_Single($my_header, 'colspan="' . MSRE_COLUMNS . '" class="header"');
     }
 
-    // write out the table header(s)
-    TRH_Single("Current contents of <b>$short_filename</b>:", 'colspan="' . MSRE_COLUMNS . '"');
-    // display the file contents
+    // Write out the table header(s)
+    TRH_Single(sprintf(__('contentsof55'), $short_filename), 'colspan="' . MSRE_COLUMNS . '"');
```
Comment (Skywalker-11) [inline mailscanner/msre_edit.php]: `contentsof55` is undefined

File: mailscanner/msre_edit.php [hunk 25]
```diff
@@ -291,112 +257,115 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
-        // now continue on..
-        $rule_action_select_html .= ">\n" .
-            "  <option value=\"\" selected>----</option>\n" .
-            "  <option value=\"Delete\">Delete</option>\n" .
-            $rule_action_select_options . "</select>\n";
+        
+        // Now continue on.
+        $rule_action_select_html .= '>' . "\n" .
+            '  <option value="" selected>----</option>' . "\n" .
+            '  <option value="Delete">' . __('delete55') . '</option>' . "\n" .
```
Comment (Skywalker-11) [inline mailscanner/msre_edit.php]: localization is missing for `delete55`

Comment (Skywalker-11) [inline mailscanner/msre_edit.php]: de would be `Löschen`

File: mailscanner/msre_edit.php [hunk 26]
```diff
@@ -457,90 +427,87 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
-        // something new for v0.2.1 ... I'm going to try to put a box
+        // Something new for v0.2.1. I'm going to try to put a box
         // around each ruleset, so it's easier to pick them out from
-        // each other in the list...
-        echo "<tr>\n" .
-            "<td class=\"$boxclass\" colspan=\"" . MSRE_COLUMNS . "\">\n" .
-            "<table border=\"0\">\n";
+        // each other in the list.
+        echo '<tr>' . "\n" .
+            '<td class="' . $boxclass . '" colspan="' . MSRE_COLUMNS . '>' . "\n" . '<table border="0">' . "\n";
```
Comment (Skywalker-11) [inline mailscanner/msre_edit.php]: here's the culprit:
`MSRE_COLUMNS . '">'`
the quote is missing

File: mailscanner/msre_edit.php (hunk 24 above)
Comment (stefaweb) [inline mailscanner/msre_edit.php]: A way to have the <b> inside the code?
(not in lang file)

**Comment by stefaweb** [comment]
I still have the submit button at top left. This is present in the original code!

e6adff50-1bc1-11e7-8470-2d8fa4bd035c

File: mailscanner/msre_edit.php (hunk 24 above)
Comment (stefaweb) [inline mailscanner/msre_edit.php]: done

File: mailscanner/msre_edit.php (hunk 25 above)
Comment (stefaweb) [inline mailscanner/msre_edit.php]: done

File: mailscanner/msre_edit.php (hunk 26 above)
Comment (stefaweb) [inline mailscanner/msre_edit.php]: done

**Comment by stefaweb** [comment]
Something not normal.

When I'm on msre_index.php, I can see this in the URL:

`https://server/msre_index.php`

If I click on a rule and go in "edit mode", I see this in the URL:

`https://server/msre_edit.php?token=b3a9a30aa70089bd1289fdd4690c76ef0246f2acf27e71024f5d558dc8473a23&file=bounce.rules`

If I go back to msre_index.php or other.php with the links, I no longer see the token part.

**Comment by Skywalker-11** [comment]
it is only there where the code depends on GET eg. normal links. Everywhere where we can use POST its "hidden" in the form

**Comment by stefaweb** [comment]
Hello!

The localization is over.

The only problem still here is the "submit" button at top left of the screen.

`echo '<input type="submit" name="submit" value="' . __('submit55') . '">' . "\n";`

I suspect that line 134 is useless. It appears that this line was added after the new session system was set up.

What I should do?

File: mailscanner/msre_edit.php [hunk 27]
```diff
@@ -291,112 +257,115 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
-                    // and gets a checkbox
+                    // And gets a checkbox
                     $field_name = $part_name;
                     if ($rule_disabled) {
                         $field_name .= '_disabled';
                     }
-                    $checkbox_html = "<input type=\"checkbox\" name=\"$field_name\" value=\"";
-                    $checkbox_html .= 'and';
+                    $checkbox_html = '<input type="checkbox" name="' . $field_name . '" value="';
+                    $checkbox_html .= __('and55');
```
Comment (Skywalker-11) [inline mailscanner/msre_edit.php]: I suggest to not use localized value here as it is part of the logic. The localization for it happens in l345

File: mailscanner/msre_edit.php [hunk 28]
```diff
@@ -291,112 +257,115 @@ function Show_Form($status_msg, $short_filename, $file_contents, $CONF_ruleset_k
...
                         $checkbox_html .= ' disabled ';
                     }
-                    $checkbox_html .= '>and';
+                    $checkbox_html .= '> ' . __('and55');
                     if ($rule_disabled) {
-                        $checkbox_html .= "\n<input type=\"hidden\" name=\"$part_name\" value=\"";
+                        $checkbox_html .= "\n" . '<input type="hidden" name="' . $part_name . '" value="';
                         if ($value) {
-                            $checkbox_html .= 'and';
+                            $checkbox_html .= __('and55');
```
Comment (Skywalker-11) [inline mailscanner/msre_edit.php]: Same here. Should not be localized

File: mailscanner/msre_edit.php (hunk 27 above)
Comment (stefaweb) [inline mailscanner/msre_edit.php]: Done.

File: mailscanner/msre_edit.php (hunk 28 above)
Comment (stefaweb) [inline mailscanner/msre_edit.php]: Done.

**Comment by stefaweb** [comment]
Need this for de.php:

`    'backmw55' => 'Back to Tools and Links',`

**Comment by Skywalker-11** [comment]
'backmw55' is already translated for de 😄

**Comment by endelwar** [comment]
> I don't like the system to "reload" MailScanner. Do we really needed that as MS reload itself regularly? Do we have another way to do this?

regarding reload MailScanner: should a sysadmin expect that the new rule is applied immediately after saving it? what about autoreload longer that default 15 minutes?

**Comment by stefaweb** [comment]
We can may be add  a "service mailscanner reload" directly in php code with a new entry in the /etc/sudoers.d/mailwatch ?

```
# /etc/sudoers.d/mailwatch
#
User_Alias MAILSCANNER = mailwatch
Cmnd_Alias MAILSCANLINT = /usr/sbin/MailScanner
Cmnd_Alias SPAMASSASSINLINT = /usr/bin/spamassassin
Cmnd_Alias EXIM_QUEUE_IN = /usr/sbin/exim -bpc
Cmnd_Alias EXIM_QUEUE_OUT = /usr/sbin/exim -bpc -DOUTGOING
Cmnd_Alias SENDMAIL_QUEUE_IN = /usr/sbin/sendmail -bp -OQueueDirectory=/var/spool/mqueue.in
Cmnd_Alias SENDMAIL_QUEUE_OUT = /usr/sbin/sendmail -bp
Cmnd_Alias MSRELOAD = /usr/sbin/service mailscanner reload
MAILSCANNER ALL= NOPASSWD: MAILSCANLINT, SPAMASSASSINLINT, EXIM_QUEUE_IN, EXIM_QUEUE_OUT, SENDMAIL_QUEUE_IN, SENDMAIL_QUEUE_OUT, MSRELOAD
```

**Comment by endelwar** [comment]
We should add a new entry in conf.php for "mailscanner reload command" (different os/distro use different commands), but this is food for a different issue.

If translation PR is ok I'll merge it

**Comment by Skywalker-11** [comment]
de.php `'newrule55' =>'Neue Regel hinzufügen: '`

**Comment by stefaweb** [comment]
Ok for the merge.

If someone else can handle this:
```
Functions to move in functions.php
See issue #624 for bug.
```

Have a nice evening.

**Pull request merged**